### PR TITLE
Flag to patch TE GEMMs with recalculate verification.

### PR DESCRIPTION
**Pull request by kwyss-nvidia**
Co-authored by: Cyril Meurillon <cmeurillon@nvidia.com>

# What does this PR do ?

Adds an option to verify GEMM results are repeatable.

This can identify bugs that are either faults in the determinism of the BLAS or issues with a device.
It has the potential to triage misbehaving devices.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by kwyss-nvidia** [comment]
Tagging @deepakn94

File: megatron/core/extensions/gemm_patch.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    if _GLOBAL_IS_PATCHED_ is not None and _GLOBAL_IS_PATCHED_:
+        return
+    transformer_engine.pytorch.cpp_extensions.gemm.general_gemm = new_general_gemm
+    transformer_engine.pytorch.cpp_extensions.gemm.general_grouped_gemm = new_general_grouped_gemm
+
+    rank = int(os.environ.get("RANK", 0))
+    for module in sys.modules:
+        if 'general_gemm' in dir(sys.modules[module]):
+            if rank == 0:
+                print(f"PATCHING general_gemm IN MODULE: {module}")
```
Comment (deepakn94) [inline megatron/core/extensions/gemm_patch.py]: No `print`s in `megatron/core`. Use Python's `logging` module instead.

File: megatron/core/extensions/gemm_patch.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+def patch_te_gemms() -> None:
+    global _GLOBAL_IS_PATCHED_
+    if _GLOBAL_IS_PATCHED_ is not None and _GLOBAL_IS_PATCHED_:
+        return
+    transformer_engine.pytorch.cpp_extensions.gemm.general_gemm = new_general_gemm
+    transformer_engine.pytorch.cpp_extensions.gemm.general_grouped_gemm = new_general_grouped_gemm
+
+    rank = int(os.environ.get("RANK", 0))
+    for module in sys.modules:
+        if 'general_gemm' in dir(sys.modules[module]):
```
Comment (deepakn94) [inline megatron/core/extensions/gemm_patch.py:145]: Is this the cleanest way to monkey patch this?

File: megatron/core/extensions/gemm_patch.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    elif isinstance(t, (NVFP4Tensor, NVFP4TensorStorage)):
+        return "nvfp4"
+    elif isinstance(t, torch.Tensor):
+        return DTYPES.get(t.dtype, "torch-unknown")
+    else:
+        return "unknown"
+
+
+def new_general_gemm(
+    A,
```
Comment (deepakn94) [inline megatron/core/extensions/gemm_patch.py]: Do we actually need the full argument lists here and below? Can we make this `*args, **kwargs`?

File: megatron/core/extensions/gemm_patch.py (hunk 1 above)
Comment (deepakn94) [reply megatron/core/extensions/gemm_patch.py]: You can use https://gitlab-master.nvidia.com/ADLR/megatron-lm/-/blob/main/megatron/core/utils.py?ref_type=heads#L737 to just log on a single rank.

Comment (kwyss-nvidia) [reply megatron/core/extensions/gemm_patch.py]: Done.

File: megatron/core/extensions/gemm_patch.py (hunk 2 above)
Comment (kwyss-nvidia) [reply megatron/core/extensions/gemm_patch.py:145]: I'm not sure. There are two parts to this. Future import statements should see the result of

```
    transformer_engine.pytorch.cpp_extensions.gemm.general_gemm = new_general_gemm
```

The sys.modules patching is to make sure that any existing resolutions to these names also get updated.

Perhaps it's possible to have the patch ordered before any transformer engine imports run, but that seems precarious.

Do you have an idea of a cleaner way? Monkey patching is not a very clean pattern, but this does work.

File: megatron/core/extensions/gemm_patch.py (hunk 3 above)
Comment (kwyss-nvidia) [reply megatron/core/extensions/gemm_patch.py]: This would rely on TE calls to general_gemm using keyword arguments and not positional arg resolution for keyword arguments.

TE style seems pretty consistent about it, so that looks like a reliable assumption at first glance.

Comment (kwyss-nvidia) [reply megatron/core/extensions/gemm_patch.py]: Is this what you have in mind? I will try running with this arg after these changes to check I didn't break anything.

File: megatron/training/training.py [hunk 4]
```diff
@@ -614,6 +614,10 @@ def pretrain(
     args = get_args()
     timers = get_timers()
 
+    if args.repeat_te_gemms_check:
+        from megatron.core.extensions.gemm_patch import patch_te_gemms
```
Comment (jaredcasper) [inline megatron/training/training.py:620]: Should make sure that we are using TE?

File: megatron/core/extensions/gemm_patch.py (hunk 2 above)
Comment (jaredcasper) [reply megatron/core/extensions/gemm_patch.py:145]: How confident are we there isn't another "general_gemm" in there that isn't the TE gemm?

File: megatron/core/extensions/gemm_patch.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+            A, B, workspace, out=out_arg, accumulate=accumulate, bias=bias_arg, **kwargs
+        )
+        out_res = res[0]
+        is_rerun = run_idx == 1
+        if will_rerun or is_rerun:
+            cksum[run_idx] = out_res.sum(dtype=torch.float32).item()
+        if will_rerun:
+            rerun_res = out_res
+            continue
+        if is_rerun:
```
Comment (jaredcasper) [inline megatron/core/extensions/gemm_patch.py]: Why all this will_rerun/is_rerun logic? Why not just put this stuff after the loop?

File: megatron/core/extensions/gemm_patch.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        else:
+            out_arg = out
+            bias_arg = bias
+
+        res = orig_general_gemm(
+            A, B, workspace, out=out_arg, accumulate=accumulate, bias=bias_arg, **kwargs
+        )
+        out_res = res[0]
+        is_rerun = run_idx == 1
+        if will_rerun or is_rerun:
```
Comment (jaredcasper) [inline megatron/core/extensions/gemm_patch.py]: How will this ever not be true?

File: megatron/core/extensions/gemm_patch.py (hunk 2 above)
Comment (cyme) [reply megatron/core/extensions/gemm_patch.py:145]: Re: patching before any TE imports -- AFAIK it's impossible b/c of a chicken-and-egg situation. We need to import TE in order to patch general_gemm, and importing TE triggers additional imports that expose general_gemm to other modules before we have a chance to perform the patch.

Agreed monkey patching is not a clean pattern. But I'm not aware of a clean and effective solution short of implementing this capability in TE itself. We can take that as a next step.

Re: risk of "another general_gemm in there that isn't the TE gemm" -- it's definitely a risk, at least in theory. We can protect against it by checking if the symbol we're patching resolves to the TE GEMM.

File: megatron/core/extensions/gemm_patch.py (hunk 5 above)
Comment (kwyss-nvidia) [reply megatron/core/extensions/gemm_patch.py]: These aren't referring to a coarse grain rerun. will_rerun/is_rerun are whether this is the first or second invocation of the GEMM in a loop.

File: megatron/core/extensions/gemm_patch.py (hunk 6 above)
Comment (kwyss-nvidia) [reply megatron/core/extensions/gemm_patch.py]: Good point. It can be elided.

File: megatron/core/extensions/gemm_patch.py (hunk 2 above)
Comment (kwyss-nvidia) [reply megatron/core/extensions/gemm_patch.py:145]: This is a debugging feature. How much extra complexity is warranted to check against name collisions and ensure TE use if the user has opted into a whitebox flag?

File: megatron/training/training.py (hunk 4 above)
Comment (kwyss-nvidia) [reply megatron/training/training.py:620]: As in "transformer_impl"==TE?

File: megatron/core/extensions/gemm_patch.py (hunk 5 above)
Comment (kwyss-nvidia) [reply megatron/core/extensions/gemm_patch.py]: Renamed for clarity.

File: megatron/core/extensions/gemm_patch.py (hunk 2 above)
Comment (kwyss-nvidia) [reply megatron/core/extensions/gemm_patch.py:145]: Added a check that we only patch the TE modules to increase safety.

File: megatron/training/training.py (hunk 4 above)
Comment (kwyss-nvidia) [reply megatron/training/training.py:620]: Done.

File: megatron/core/extensions/gemm_patch.py (hunk 3 above)
Comment (deepakn94) [reply megatron/core/extensions/gemm_patch.py]: Looks good, resolving.

File: megatron/core/extensions/gemm_patch.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+        res = orig_general_gemm(
+            A, B, workspace, out=out_arg, accumulate=accumulate, bias=bias_arg, **kwargs
+        )
+        out_res = res[0]
+        is_second_iter = run_idx == 1
+        cksum[run_idx] = out_res.sum(dtype=torch.float32).item()
+        if is_first_iter:
+            rerun_res = out_res
+            continue
+        if is_second_iter:
```
Comment (jaredcasper) [inline megatron/core/extensions/gemm_patch.py]: If this only runs on the second iteration, can't it just run outside of the loop after it finishes? Even with the name changes this loop logic with is_first_iter and is_second_iter is unnecessarily complex.

Comment (deepakn94) [reply megatron/core/extensions/gemm_patch.py]: Yeah, I think that's a good suggestion.

Comment (kwyss-nvidia) [reply megatron/core/extensions/gemm_patch.py]: Moving the check and return outside of the loop is equivalent, and is arguably easier to read (at least to my judgment). Thanks for the feedback.

File: megatron/core/extensions/gemm_patch.py [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+        )
+        out_res = res[0]
+        cksum[run_idx] = out_res.sum(dtype=torch.float32).item()
+        if is_first_iter:
+            rerun_res = out_res
+            continue
+
+    # Check the consistency of the first and second run and return.
+    same_cksum = math.fabs(cksum[1] - cksum[0]) / (math.fabs(cksum[1]) + EPSILON) <= CKSUM_TOLERANCE
+    if not same_cksum:
```
Comment (cyme) [inline megatron/core/extensions/gemm_patch.py]: What is the value of doing a cksum comparison if we do a bit-wise comparison anyway?

File: megatron/core/extensions/gemm_patch.py [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+        logger.log(
+            logging.ERROR,
+            f"RANK {get_rank_info()}: DIFFERENT CHECKSUM ON 2ND RUN for dtype "
+            f"{get_dtype(A)}. {cksum[1]} != {cksum[0]}, "
+            f"shapes A:{list(A.size())},B:{list(B.size())},out:{list(out_res.shape)}",
+        )
+    elif not torch.equal(rerun_res, out_res):
```
Comment (cyme) [inline megatron/core/extensions/gemm_patch.py]:
```suggestion
    elif not torch.equal(rerun_res.view(dtype=torch.uint8), out_res.view(dtype=torch.uint8))):
```
Doing the equality test in the int domain ensure a bit-wise comparison. This correctly handles NaNs, which are never equal in the FP domain.

File: megatron/core/extensions/gemm_patch.py [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+        logger.log(
+            logging.ERROR,
+            f"RANK {get_rank_info()}: DIFFERENT BITWISE on 2ND RUN for dtype "
+            f"{get_dtype(A)}. "
+            f"shapes A:{list(A.size())},B:{list(B.size())},out:{list(out_res.shape)}",
```
Comment (cyme) [inline megatron/core/extensions/gemm_patch.py]:
```suggestion
            f"RANK {get_rank_info()}: DIFFERENT BITWISE on 2ND RUN FOR OPERATION "
            f"A:({get_dtype(A)},{list(A.size())}) x B:({get_dtype(B)},{list(B.size())}) "
            f"-> OUT:({get_dtype(out_res)},{list(out_res.size())})",
```
Logging this too as it's very useful

File: megatron/core/extensions/gemm_patch.py [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+            **kwargs,
+        )
+        output = res[0]
+        if is_first_iter:
+            saved_output = [t.clone().detach() for t in output]
+            continue
+
+    # Check comparison and return
+    for tidx in range(len(output)):
+        if not torch.equal(output[tidx], saved_output[tidx]):
```
Comment (cyme) [inline megatron/core/extensions/gemm_patch.py]:
```suggestion
        if not torch.equal(output[tidx].view(dtype=torch.uint8), saved_output[tidx].view(dtype=torch.uint8)):
```
Same suggestion as above

File: megatron/core/extensions/gemm_patch.py [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+            logging.log(
+                logging.ERROR,
+                f"RANK {get_rank_info()}: DIFFERENT GROUPED_GEMM RESULT ON 2ND RUN FOR OP "
+                f"GroupedGEMM group {tidx} = {get_dtype(A[tidx])}({list(A[tidx].size())}) "
+                f"x {get_dtype(B[tidx])}({list(B[tidx].size())}) -> "
+                f"{get_dtype(output[tidx])}({list(output[tidx].size())})",
```
Comment (cyme) [inline megatron/core/extensions/gemm_patch.py]:
```suggestion
                f"RANK {get_rank_info()}: DIFFERENT GROUPED_GEMM RESULT ON 2ND RUN FOR "
                f"GROUP {tidx}, A:({get_dtype(A[tidx])},{list(A[tidx].size())}) "
                f"x B:({get_dtype(B[tidx])},{list(B[tidx].size())}) -> "
                f"OUT: ({get_dtype(output[tidx])},{list(output[tidx].size())})",
```
nit: made logging consistent with `general_gemm()` for easier parsing

File: megatron/core/extensions/gemm_patch.py [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+def new_general_gemm(A, B, workspace, out=None, accumulate=False, bias=None, **kwargs):
+    """
+    A wrapper for general_gemm that calls the original method twice
+    and compares results for consistency.
+    """
+    cksum = [0, 0]
+    # Run twice for comparison.
+    for run_idx in range(2):
+        is_first_iter = run_idx == 0
+        need_output_copy = out is not None and (bias is out or accumulate)
```
Comment (cyme) [inline megatron/core/extensions/gemm_patch.py:56]: I'm not sure `bias is out` is a legitimate use case, but it doesn't hurt to check

File: megatron/core/extensions/gemm_patch.py (hunk 9 above)
Comment (cyme) [reply megatron/core/extensions/gemm_patch.py]: Alternatively to making a copy of the first result and doing an actual bit-wise comparison, we can use a hash. This would reduce a bit the overhead. PyTorch2.9 introduced a `hash_tensor()` function. On previous versions, a possible poor man hash is a sum of the tensor viewed as int64s.

File: megatron/core/extensions/gemm_patch.py (hunk 8 above)
Comment (kwyss-nvidia) [reply megatron/core/extensions/gemm_patch.py]: It provides a bit of information whether the error was large in magnitude.

Comment (kwyss-nvidia) [reply megatron/core/extensions/gemm_patch.py]: Arguably marginal value. Removed.

File: megatron/core/extensions/gemm_patch.py (hunk 13 above)
Comment (kwyss-nvidia) [reply megatron/core/extensions/gemm_patch.py:56]: Yep. Worth handling the edge case. Are you requesting a change?

File: megatron/core/extensions/gemm_patch.py (hunk 11 above)
Comment (kwyss-nvidia) [reply megatron/core/extensions/gemm_patch.py]: Done.

**Comment by Phlip79** [comment]
**We are changing our review process and marking all open, unlabeled PRs as draft. This change will go in effect starting once https://github.com/NVIDIA/Megatron-LM/pull/3659 is merged.**

Moving forward, all PRs will be required to start as draft PRs. If you wish to get your PR merged, mark your PR as “Ready for review”. Read more about the new process at [submit.md](https://github.com/Phlip79/Megatron-LM/blob/d3763629bcdd04acca52477af7eb8795090426ff/docs/developer/submit.md).
